hashintel/hash
# H-1346: Seed existing linear types via migrations instead of type initializers + fix the existing linear integration

## 🌟 What is the purpose of this PR?




This PR adds the existing `User`, `Organization` and `Issue` entity types to the migration scripts, removing the need for type initializers. As part of this, it adds the linear types to the `generate-ontology-type-ids` script so those become available statically.

It also fixes the existing implementation of the linear integration, by:
- updating redirects that were still using `/account/integrations/...` instead of `/settings/integrations/...`
- updating the linear temporal workflow to use the `systemAccountId` instead of the `publicUserAccountId` when trying to read/update/create entities (addresses H-921)
- updating the `syncLinearIntegrationWithWorkspaces` resolver to add the `systemAccountId` as a member of the org, when syncing with an org's web. When syncing with a user's web, an error is thrown as it is currently not possible to give it access to that web. Additionally, the `systemAccountId` is given view access to the `LinearIntegration` and `LinearUserSecret` entities, so that these can be retrieved via the `systemAccountId` in the sync-back watcher.
- updating the seed script to make the `Alice` user the owner of the `Example` org, so that `Alice` has permission to add the `systemAccountId` as a member of the `Example` org's account group.
- when updating an entity in the linear temporal workflow via a webhook, merge the properties of the updated entity with the properties of the prior entity to prevent validation errors as some of the required properties aren't included in the data the webhook receives
- re-enables redis, and gracefully handles when the `systemAccountId` cannot view an entity because it lacks permissions.

## 🔗 Related links





- H-1346

## 🔍 What does this change?




See description and commits

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



- It's not possible to setup a linear integration for a user's web, as we don't have a way of giving the `systemAccountId` access to that web
- When a linear integration is removed, we need to somehow determine whether another linear integration is associated with that org's web before revoking access, otherwise we break other user's linear integrations with that web
- Only web owners can create a linear integration for an org's web, as they require the permission to add a member to the account group
- In the linear sync-back watcher we currently assume the entity is in a org's web, not a user's web.

## 🐾 Next steps



- Improve the temporal workflows via a statically declared mapping between the linear data structure and the ontology types we have in the graph. This should also handle updating of links between the entities (e.g. a new assignee to an issue) (H-1347)
- Seed the remaining types we want to showcase via the migrations
- Fix syncing changes back into linear

## 🛡 What tests cover this?



None

## ❓ How to test this?



1. Checkout the branch locally and run the HASH App on a fresh database, with the required linear environment variables
2. Check the linear types have seeded
3. It should also be possible to link a linear account to a web, and seed some entities. This flow can be initiated at `/settings/integrations`, and will the `LINEAR_CLIENT_ID` and `LINEAR_CLIENT_SECRET` to be set.
4. The linear webhook should also work as it did previously, which requires the `LINEAR_WEBHOOK_SECRET` to be set and the linear app to point to your localhost via a `ngrok` port-forward setup (I used `ngrok http 5001` to set this up)
5. The sync-back watcher should work, by ensuring redis is enabled. Update a linear entity, and the change should be pushed to linear.

## 📹 Demo



https://github.com/hashintel/hash/assets/42802102/26a5476c-96d1-4185-83fd-dfd1853e63d3
